runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PTransformTranslation.java

```diff
@@ -98,6 +106,29 @@
     return ImmutableMap.copyOf(translators);
   }
 
+  private static Map<String, TransformPayloadTranslator> loadTransformRehydrators() {
+    HashMap<String, TransformPayloadTranslator> rehydrators = new HashMap<>();
+
+    for (TransformPayloadTranslatorRegistrar registrar :
+        ServiceLoader.load(TransformPayloadTranslatorRegistrar.class)) {
+
+      Map<String, ? extends TransformPayloadTranslator> newRehydrators =
+          registrar.getTransformRehydrators();
+
+      Set<String> alreadyRegistered =
+          Sets.intersection(rehydrators.keySet(), newRehydrators.keySet());
+
+      if (!alreadyRegistered.isEmpty()) {
+        throw new IllegalArgumentException(
```

Comment:
Filed https://issues.apache.org/jira/browse/BEAM-3069. It will require some `equals` methods I think, to get right. Since we currently don't have a double-shaded service loader issue, I'd like to merge-then-fix rather than fix-then-merge.